File: tools/postcost.js
/* WHAT THE POST CHAIN COSTS PER FRAME. Driven by tools/post_cost.py.
 *
 * The house rules for this measurement are hud_cost.py's and they are not
 * optional, because getting them wrong turns the instrument into the result:
 *
 *   - N draws inside ONE rAF, not one draw per rAF. Per-frame rAF measures the
 *     display's cadence, not the work.
 *   - Force the raster ONCE, at the end. A getImageData (or a readPixels)
 *     every iteration can demote an accelerated canvas to software, and then
 *     the number is of a different renderer than the one that ships.
 *   - m.shake = 0. The camera offset calls Math.random(), so leaving it on
 *     makes every iteration draw a slightly different frame.
 *
 * And the rule this file adds: REPORT THE RENDERER. A number off SwiftShader
 * is not Rick's machine and is not a phone. What transfers between runtimes is
 * the RATIO between rows measured back to back in one session, which is the
 * only question the chain has to answer -- did it make the frame materially
 * more expensive than the 2D draw already was?
 */
(cfg) => {
  AC.setResolution(cfg.w, cfg.h);
  const m = new AC.Match(cfg.a, cfg.b, cfg.seed >>> 0);
  m.introT = 0;
  AC.__inject(m);
  AC.SFX.play = function () {};
  AC.SFX.resume = function () {};
  const dt = AC.CONFIG.physics.dt;
  while (m.t < cfg.at) m.step(dt);
  m.shake = 0;

  const cv = document.getElementById('cv');
  const ctx = cv.getContext('2d');
  const one = new Uint8Array(4);

  const R = AC.renderer;
  const state = () => ({
    enabled: true, dt: 1 / 60,
    rect: { x: R.pad * R.k, y: R.arenaTop * R.k, w: R.aw * R.k, h: R.ah * R.k },
    cine: AC.CINE ? { on: !!AC.CINE.on, cut: !!AC.CINE.cut } : null,
  });

  const gl = (typeof POSTFX !== 'undefined' && POSTFX.gl)
    ? POSTFX.gl.getContext('webgl2') : null;
  const dbg = gl && gl.getExtension('WEBGL_debug_renderer_info');
  const renderer = !gl ? 'no webgl'
    : (dbg ? gl.getParameter(dbg.UNMASKED_RENDERER_WEBGL)
           : gl.getParameter(gl.RENDERER));

  /* MEASURES THE BUILD'S OWN CHAIN, not a copy of it. postcost used to
     create its own Post instance and time AC.__draw plus post.render -- which
     was right while the chain lived only in the app, and became wrong the
     moment post_build.py put it in the build: AC.__draw now composites by
     itself, so that arrangement was timing the work TWICE.

     POSTFX.on is the honest switch. Off is the shipping renderer with the
     chain off; on is everything, three draws and all. */
  /* BARE NAME, NOT window.POSTFX. POSTFX is a top-level const in the build,
     which makes it a lexical global and NOT a property of window -- the same
     trap CINE and `renderer` set, now for the third time today. It resolves
     by name here because this runs in the game's own realm. */
  const FX = (typeof POSTFX !== 'undefined') ? POSTFX : null;
  if (!FX) return { err: 'this build has no POSTFX -- run post_build.py' };
  FX.init();

  const CONFIGS = [
    { name: '2D draw only', on: false },
    { name: '+ chain, no effects', on: true },
    { name: '+ bloom ' + SWBPost.SPREAD.DEFAULT, on: true,
      bloom: SWBPost.SPREAD[SWBPost.SPREAD.DEFAULT] },
    { name: '+ trails ' + SWBPost.TRAILS.DEFAULT, on: true,
      trails: SWBPost.TRAILS[SWBPost.TRAILS.DEFAULT] },
    { name: '+ grade ' + SWBPost.GRADE.DEFAULT, on: true,
      grade: SWBPost.GRADE[SWBPost.GRADE.DEFAULT] },
    { name: '+ ALL (as chosen)', on: true,
      bloom: SWBPost.SPREAD[SWBPost.SPREAD.DEFAULT],
      trails: SWBPost.TRAILS[SWBPost.TRAILS.DEFAULT],
      grade: SWBPost.GRADE[SWBPost.GRADE.DEFAULT] },
  ];

  const measure = (c) => new Promise((res) => {
    FX.post.setBloom(c.bloom || null);
    FX.post.setTrails(c.trails || null);
    FX.post.setGrade(c.grade || null);
    FX.on = c.on;
    FX.reset();
    for (let i = 0; i < cfg.warm; i++) AC.__draw(m);
    requestAnimationFrame(() => {
      const t0 = performance.now();
      for (let i = 0; i < cfg.n; i++) AC.__draw(m);
      /* One sync at the end, never per iteration -- a readback every frame can
         demote an accelerated canvas to software and time a different
         renderer than the one that ships. */
      if (c.on) gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, one);
      else ctx.getImageData(0, 0, 1, 1);
      res((performance.now() - t0) / cfg.n);
    });
  });

  return (async () => {
    const rows = [];
    for (let rep = 0; rep < cfg.reps; rep++) {
      for (let i = 0; i < CONFIGS.length; i++) {
        const ms = await measure(CONFIGS[i]);
        rows.push({ name: CONFIGS[i].name, rep: rep, ms: ms });
      }
    }
    FX.post.setBloom(SWBPost.SPREAD[SWBPost.SPREAD.DEFAULT]);
    FX.post.setTrails(SWBPost.TRAILS[SWBPost.TRAILS.DEFAULT]);
    FX.post.setGrade(SWBPost.GRADE[SWBPost.GRADE.DEFAULT]);
    FX.on = true;
    return { renderer: renderer, size: cv.width + 'x' + cv.height,
             at: +m.t.toFixed(2), rows: rows };
  })();
}

File: tools/post_cost.py
#!/usr/bin/env python3
"""WHAT DOES THE POST CHAIN COST PER FRAME?

    python post_cost.py                 on the real GPU, through Electron
    python post_cost.py --headless      through Playwright, i.e. SwiftShader

`docs/RENDERER-BRIEF.md` §7 gate 3: frame cost measured, not felt. "A post
chain that adds 8 ms is not free at 120 fps -- it is the whole budget."

WHICH RUNTIME, AND WHY IT MATTERS MORE HERE THAN ANYWHERE ELSE IN THIS REPO.
Everywhere else the two Chromiums are interchangeable because they agree to
the last bit (docs/RUNTIME-DRIFT.md). Not for this. Playwright launches with
--disable-gpu, so its WebGL is SwiftShader: a software rasteriser, and a
number off it is a measurement of SwiftShader. Electron uses the machine's own
GPU, which is what Rick watches on. So this defaults to Electron and always
prints the renderer it measured.

WHAT TRANSFERS is the RATIO between rows measured back to back in one session
-- hud_cost.py's point, and the only question the chain has to answer: did it
make the frame materially more expensive than the 2D draw already was?

The median of `--reps` repetitions is reported, not the mean: one scheduling
hiccup in a run of five should not become the answer.
"""
from __future__ import annotations

import argparse
import json
import pathlib
import statistics
import subprocess
import sys

HERE = pathlib.Path(__file__).resolve().parent
REPO = HERE.parent
BUILD = REPO / "02-chain" / "sc-paradox-hold-clamp.html"
POST_JS = REPO / "src" / "render" / "post.js"
COST_JS = HERE / "postcost.js"
EL_JS = HERE / "postcost_electron.js"
EL_BIN = [REPO / "app" / "node_modules" / ".bin" / n
          for n in ("electron.cmd", "electron")]


def run_electron(game: pathlib.Path, cfg: dict) -> dict | None:
    exe = next((p for p in EL_BIN if p.exists()), None)
    if exe is None:
        print("! no Electron in app/node_modules -- run `npm install` in app/,")
        print("  or pass --headless and read the caveat in the header.")
        return None
    out = subprocess.run(
        [str(exe), str(EL_JS), "--game", str(game), "--cfg", json.dumps(cfg)],
        capture_output=True, text=True)
    if out.returncode != 0 or "{" not in out.stdout:
        print("! electron run failed")
        print((out.stderr or out.stdout)[-1200:])
        return None
    return json.loads(out.stdout[out.stdout.index("{"):])


def run_headless(game: pathlib.Path, cfg: dict) -> dict | None:
    from scpage import game as page_ctx
    with page_ctx(game_path=game) as (page, errors):
        page.add_script_tag(content=POST_JS.read_text(encoding="utf-8"))
        page.evaluate("() => { window.__frozen = true; }")
        out = page.evaluate("(" + COST_JS.read_text(encoding="utf-8") + ")", cfg)
        if errors:
            print("! page errors:")
            for e in errors[:6]:
                print("   ", e)
            return None
    return out


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--game", default=str(BUILD))
    ap.add_argument("--headless", action="store_true",
                    help="Playwright instead of Electron. SOFTWARE rasteriser.")
    ap.add_argument("--a", default="paradox")
    ap.add_argument("--b", default="heartwood")
    ap.add_argument("--seed", type=int, default=25064)
    ap.add_argument("--at", type=float, default=22.0,
                    help="seconds of fight before measuring -- seals up, "
                         "statuses live, cracks grown")
    ap.add_argument("--n", type=int, default=60, help="draws per repetition")
    ap.add_argument("--warm", type=int, default=12)
    ap.add_argument("--reps", type=int, default=5)
    ap.add_argument("--w", type=int, default=1080)
    ap.add_argument("--h", type=int, default=1920)
    ap.add_argument("--json", metavar="PATH")
    A = ap.parse_args()

    game = pathlib.Path(A.game).resolve()
    if not game.exists():
        print(f"! {game} does not exist")
        return 2
    for f in (POST_JS, COST_JS):
        if not f.exists():
            print(f"! {f} does not exist")
            return 2

    cfg = {"a": A.a, "b": A.b, "seed": A.seed, "at": A.at, "n": A.n,
           "warm": A.warm, "reps": A.reps, "w": A.w, "h": A.h}

    print(f"\nPOST COST  {A.a} v {A.b} seed {A.seed} at t={A.at}s  "
          f"{A.w}x{A.h}  {A.n} draws x {A.reps} reps")
    out = (run_headless(game, cfg) if A.headless else run_electron(game, cfg))
    if out is None:
        return 1

    soft = any(s in out["renderer"] for s in ("SwiftShader", "llvmpipe", "Software"))
    print(f"  {out['renderer']}")
    if soft:
        print("  ^ SOFTWARE RASTERISER. These are not Rick's numbers and not a")
        print("    phone's. Read the RATIO column and ignore the milliseconds.")

    by = {}
    order = []
    for r in out["rows"]:
        if r["name"] not in by:
            by[r["name"]] = []
            order.append(r["name"])
        by[r["name"]].append(r["ms"])

    base = statistics.median(by[order[0]])
    print(f"\n  {'configuration':<24} {'ms/frame':>9} {'vs 2D':>8} "
          f"{'added':>9}  spread")
    for name in order:
        v = sorted(by[name])
        med = statistics.median(v)
        print(f"  {name:<24} {med:>9.3f} {med / base:>7.2f}x "
              f"{med - base:>+8.3f}  {v[0]:.2f}..{v[-1]:.2f}")

    chain = statistics.median(by[order[1]]) - base
    both = statistics.median(by[order[-1]]) - base
    print(f"\n  the chain with nothing on costs {chain:+.3f} ms -- that is the"
          f" upload,\n  two copies and the readback, and it is the floor.")
    print(f"  the whole chosen chain costs {both:+.3f} ms.")
    budget60, budget120 = 16.67, 8.33
    print(f"\n  a frame has {budget60:.2f} ms at 60fps and {budget120:.2f} at 120.")
    print(f"  total as chosen: {statistics.median(by[order[-1]]):.2f} ms "
          f"= {100 * statistics.median(by[order[-1]]) / budget60:.0f}% of the 60fps"
          f" budget, {100 * statistics.median(by[order[-1]]) / budget120:.0f}% of 120.")
    if soft:
        print("  (on a software rasteriser, so those percentages are of a"
              " budget\n   this machine was never going to meet. The ratio is"
              " the finding.)")

    if A.json:
        pathlib.Path(A.json).write_text(json.dumps(out, indent=1))
        print(f"\n  wrote {A.json}")
    print()
    return 0


if __name__ == "__main__":
    sys.exit(main())
